File: seed.rb
# Idempotent seed for the Summit Roofing Co. demo tenant on self-hosted Chatwoot.
# Run via:  rails runner /tmp/summit_support_seed.rb   (inside the chatwoot rails container)
#
# Creates / updates (all by stable natural keys, safe to re-run):
#   * a super-admin User (the agent),
#   * the Account "Summit Roofing Co.",
#   * an API channel inbox,
#   * 2 contacts,
#   * ~7 roofing-support conversations across open / pending / resolved,
#   * an inbound message on each so the queue/feed has real subjects.
#
# Prints on success:
#   SEED_OK account=<id> inbox=<id> contacts=<n> conversations=<n> open=<n> ...
#   ACCOUNT_ID=<id>
#   ACCESS_TOKEN=<user.access_token.token>

ADMIN_EMAIL    = 'admin@summitroofing.test'
ADMIN_NAME     = 'Summit Support Agent'
ADMIN_PASSWORD = ENV.fetch('CHATWOOT_ADMIN_PASSWORD', 'replace-me-password')
ACCOUNT_NAME   = 'Summit Roofing Co.'
INBOX_NAME     = 'Summit Roofing Support'

# --- 1. super-admin user (the agent we hand an access token to) --------------
user = User.find_by(email: ADMIN_EMAIL)
unless user
  user = User.new(
    name: ADMIN_NAME,
    email: ADMIN_EMAIL,
    password: ADMIN_PASSWORD,
    password_confirmation: ADMIN_PASSWORD
  )
  user.confirmed_at = Time.current
  user.skip_confirmation! if user.respond_to?(:skip_confirmation!)
  user.save!
end
# Promote to super admin (platform-level). Idempotent.
user.update!(type: 'SuperAdmin') unless user.type == 'SuperAdmin'

# --- 2. account --------------------------------------------------------------
account = Account.find_by(name: ACCOUNT_NAME) || Account.create!(name: ACCOUNT_NAME, locale: 'en')

# Link the user to the account as administrator (the agent identity for the API).
AccountUser.find_or_create_by!(account: account, user: user) do |au|
  au.role = :administrator
end
# Ensure role is administrator even if the row pre-existed.
au = AccountUser.find_by(account: account, user: user)
au.update!(role: :administrator) unless au.administrator?

# --- 3. API channel inbox ----------------------------------------------------
channel = Channel::Api.find_by(account: account) ||
          Channel::Api.create!(account: account, webhook_url: '')
inbox = Inbox.find_by(account: account, name: INBOX_NAME)
unless inbox
  inbox = Inbox.create!(account: account, name: INBOX_NAME, channel: channel)
end
# Make sure the agent is a member of the inbox so it can be assigned.
InboxMember.find_or_create_by!(inbox: inbox, user: user)

# --- 4. contacts -------------------------------------------------------------
def upsert_contact(account, name, email, phone)
  c = Contact.find_by(account: account, email: email)
  c ||= Contact.create!(account: account, name: name, email: email, phone_number: phone)
  c
end

henderson = upsert_contact(account, 'Dana Henderson', 'dana.henderson@example.com', '+15125550142')
maple     = upsert_contact(account, 'Marcus Webb',     'marcus.webb@example.com',    '+15125550199')

# --- 5. conversations (idempotent by a deterministic identifier) -------------
# Each entry: contact, status, priority, channel-ish source label, inbound text.
TICKETS = [
  { key: 'quote-2200-reroof', contact: :henderson, status: :open,     priority: :medium,
    source: 'Website',
    body: "Hi — I'd like a quote for a full re-roof on my 2,200 sqft house in Henderson. Existing is old asphalt shingle. What's your availability and ballpark cost?" },
  { key: 'reschedule-rain',   contact: :maple,     status: :open,     priority: :low,
    source: 'Phone',
    body: "It's supposed to rain Tuesday when your crew is scheduled. Can we push the install to Thursday instead?" },
  { key: 'warranty-ridge-leak', contact: :henderson, status: :pending, priority: :high,
    source: 'Email',
    body: "We're seeing a small leak near the ridge after last week's storm — the Oak Park job you did in spring. I believe this is under warranty. Can someone come look?" },
  { key: 'invoice-question',  contact: :maple,     status: :pending,  priority: :medium,
    source: 'Email',
    body: "Quick question on invoice #1048 — there's a $300 line item for 'tear-off disposal' I didn't expect. Can you explain what that covers?" },
  { key: 'new-roof-estimate', contact: :maple,     status: :open,     priority: :medium,
    source: 'Website',
    body: "Building a new detached garage (24x24) and need a roof on it. Can you do new construction, and how do I get an estimate?" },
  { key: 'emergency-storm',   contact: :henderson, status: :open,     priority: :urgent,
    source: 'Phone',
    body: "EMERGENCY — last night's storm tore shingles off and there's water coming into the upstairs bedroom. We need someone out today if possible. Please call ASAP." },
  { key: 'gutter-guards',     contact: :maple,     status: :resolved, priority: :low,
    source: 'Facebook',
    body: "Do you install gutter guards? If so what brands and roughly what does it run?" },
]

contacts = { henderson: henderson, maple: maple }

def status_int(status)
  Conversation.statuses[status.to_s]
end

created = 0
TICKETS.each do |t|
  contact = contacts[t[:contact]]
  ci = ContactInbox.find_by(inbox: inbox, source_id: t[:key])
  ci ||= ContactInbox.create!(inbox: inbox, contact: contact, source_id: t[:key])

  conv = Conversation.find_by(account: account, inbox: inbox, contact_inbox: ci)
  unless conv
    conv = Conversation.create!(
      account: account,
      inbox: inbox,
      contact: contact,
      contact_inbox: ci,
      additional_attributes: { 'source' => t[:source], 'ticket_key' => t[:key] }
    )
    # Seed the inbound customer message (this becomes the conversation subject/preview).
    Message.create!(
      account: account,
      inbox: inbox,
      conversation: conv,
      message_type: :incoming,
      content: t[:body],
      sender: contact
    )
    created += 1
  end

  # Normalize status + priority every run (idempotent).
  conv.update_columns(status: status_int(t[:status])) unless conv.status == t[:status].to_s
  conv.update!(priority: t[:priority]) unless conv.priority == t[:priority].to_s
  # Assign the emergency + warranty (high/urgent) to our agent so escalation reads true.
  if [:urgent, :high].include?(t[:priority]) && conv.assignee_id.nil?
    conv.update!(assignee: user)
  end
end

# --- 6. access token (the API credential the app.py uses) --------------------
# Chatwoot auto-creates an AccessToken for each User via a callback. Read it back.
token_rec = user.access_token || AccessToken.find_by(owner: user)
token_rec ||= AccessToken.create!(owner: user)
token = token_rec.token

counts = Conversation.where(account: account, inbox: inbox).group(:status).count
puts "SEED_OK account=#{account.id} inbox=#{inbox.id} contacts=#{Contact.where(account: account).count} " \
     "conversations=#{Conversation.where(account: account, inbox: inbox).count} " \
     "open=#{counts['open'] || 0} pending=#{counts['pending'] || 0} resolved=#{counts['resolved'] || 0} new=#{created}"
puts "ACCOUNT_ID=#{account.id}"
puts "ACCESS_TOKEN=#{token}"

File: seed.py
#!/usr/bin/env python3
"""Repeatable seeder for the Summit Roofing Co. demo tenant on self-hosted Chatwoot.

Bootstrap method (the reliable one for self-hosted Chatwoot, mirroring the Lago
reference): run a Ruby script via `rails runner` inside the Chatwoot rails container.
That creates/updates — idempotently — a super-admin user, the account
"Summit Roofing Co.", an API inbox, a couple of contacts, and ~6-8 roofing-support
conversations across open / pending / resolved.

Chatwoot's API needs an agent **access token** (the `access_token.token` on a User).
This seed reads that generated token back and prints `ACCESS_TOKEN=<value>` +
`ACCOUNT_ID=<id>`; we capture both and write agents/support/.env so app.py picks them
up with no manual copy/paste.

Usage:
    python3 seed.py
    CHATWOOT_CONTAINER=agentic-chatwoot-rails-1 python3 seed.py

Env knobs:
    CHATWOOT_CONTAINER  docker container of the Chatwoot rails app
                        (default: agentic-chatwoot-rails-1)
    CHATWOOT_API_URL    REST base used by app.py (default: http://localhost:3003)
    CHATWOOT_FRONT_URL  Chatwoot UI link baked into .env (default: http://192.168.40.8:3003)
"""
from __future__ import annotations

import os
import re
import subprocess
import sys
from pathlib import Path

HERE = Path(__file__).resolve().parent
SEED_RB = HERE / "seed.rb"
ENV_OUT = HERE / ".env"

CONTAINER = os.environ.get("CHATWOOT_CONTAINER", "agentic-chatwoot-rails-1")
CHATWOOT_API_URL = os.environ.get("CHATWOOT_API_URL", "http://localhost:3003")
CHATWOOT_FRONT_URL = os.environ.get("CHATWOOT_FRONT_URL", "http://192.168.40.8:3003")

# `sudo` is required to talk to the docker socket on this host.
DOCKER = ["sudo", "docker"]
IN_CONTAINER_PATH = "/tmp/summit_support_seed.rb"


def run(cmd: list[str], **kw) -> subprocess.CompletedProcess:
    return subprocess.run(cmd, text=True, capture_output=True, **kw)


def main() -> int:
    if not SEED_RB.exists():
        print(f"seed.rb not found at {SEED_RB}", file=sys.stderr)
        return 1

    # 1. Copy the Ruby seed into the running Chatwoot rails container.
    cp = run(DOCKER + ["cp", str(SEED_RB), f"{CONTAINER}:{IN_CONTAINER_PATH}"])
    if cp.returncode != 0:
        print("docker cp failed:\n" + cp.stderr, file=sys.stderr)
        return 1

    # 2. Run it with rails runner (the Chatwoot rails app).
    res = run(DOCKER + ["exec", CONTAINER, "bundle", "exec", "rails", "runner", IN_CONTAINER_PATH])
    out = res.stdout + "\n" + res.stderr

    seed_ok = re.search(r"^SEED_OK .*$", out, re.MULTILINE)
    tok_match = re.search(r"^ACCESS_TOKEN=(\S+)$", out, re.MULTILINE)
    acct_match = re.search(r"^ACCOUNT_ID=(\S+)$", out, re.MULTILINE)
    if not (seed_ok and tok_match and acct_match):
        print("Seeding did not report success. Output:\n" + out, file=sys.stderr)
        return 1

    token = tok_match.group(1)
    account_id = acct_match.group(1)
    print(seed_ok.group(0))
    print(f"ACCOUNT_ID={account_id}")
    print(f"ACCESS_TOKEN={token}")

    # 3. Persist the env so app.py picks up the live token + account automatically.
    ENV_OUT.write_text(
        f"CHATWOOT_API_URL={CHATWOOT_API_URL}\n"
        f"CHATWOOT_API_TOKEN={token}\n"
        f"CHATWOOT_ACCOUNT_ID={account_id}\n"
        f"CHATWOOT_FRONT_URL={CHATWOOT_FRONT_URL}\n"
    )
    print(f"Wrote {ENV_OUT} (CHATWOOT_API_URL, CHATWOOT_API_TOKEN, CHATWOOT_ACCOUNT_ID, CHATWOOT_FRONT_URL)")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
